src/receive.rs

```diff
@@ -1,136 +1,161 @@
-        let discarded_bytes = self.read_buffer.bytes[..self.seen_bytes].into();
+        let discarded_bytes = self.read_buffer[..self.seen_bytes].into();
         self.finish_message();
         discarded_bytes
     }
 
-    pub fn discard_all_bytes(&mut self) -> Box<[u8]> {
-        self.seen_bytes = self.read_buffer.bytes.len();
-        self.discard_message()
-    }
-
-    pub async fn progress(&mut self, stream: &mut AnyStream) -> Result<ReceiveEvent<C>, ReadError>
+    pub fn progress(&mut self) -> Result<ReceiveEvent<C>, FlowInterrupt<ReceiveError<C>>>
     where
         C: Decoder,
         for<'a> C::Message<'a>: IntoBoundedStatic<Static = C::Message<'static>>,
         for<'a> C::Error<'a>: IntoBoundedStatic<Static = C::Error<'static>>,
     {
-        loop {
+        let event = loop {
             match self.next_fragment {
                 NextFragment::Line { seen_bytes_in_line } => {
-                    if let Some(event) = self.progress_line(stream, seen_bytes_in_line).await? {
-                        return Ok(event);
-                    }
+                    break self.progress_line(seen_bytes_in_line)?;
                 }
                 NextFragment::Literal { length } => {
-                    self.progress_literal(stream, length).await?;
+                    self.progress_literal(length)?;
                 }
             };
-        }
+        };
+
+        Ok(event)
     }
 
-    async fn progress_line(
+    fn progress_line(
         &mut self,
-        stream: &mut AnyStream,
         seen_bytes_in_line: usize,
-    ) -> Result<Option<ReceiveEvent<C>>, ReadError>
+    ) -> Result<ReceiveEvent<C>, FlowInterrupt<ReceiveError<C>>>
     where
         C: Decoder,
         for<'a> C::Message<'a>: IntoBoundedStatic<Static = C::Message<'static>>,
         for<'a> C::Error<'a>: IntoBoundedStatic<Static = C::Error<'static>>,
     {
+        let max_readable_bytes = self.max_readable_bytes();
+
         let Some(crlf_result) = find_crlf(
-            &self.read_buffer.bytes[self.seen_bytes..],
+            &self.read_buffer[self.seen_bytes..max_readable_bytes],
             seen_bytes_in_line,
             self.crlf_relaxed,
         ) else {
             // No full line received yet, more data needed.
 
             // Mark the bytes of the partial line as seen.
-            let seen_bytes_in_line = self.read_buffer.bytes.len() - self.seen_bytes;
+            let seen_bytes_in_line = self.read_buffer.len() - self.seen_bytes;
             self.next_fragment = NextFragment::Line { seen_bytes_in_line };
 
-            // Read more data.
-            stream.read(&mut self.read_buffer).await?;
+            // Abort if we can't request more data.
+            if Some(max_readable_bytes) == self.max_message_size.map(|size| size as usize) {
+                self.seen_bytes = max_readable_bytes;
+                return Err(FlowInterrupt::Error(ReceiveError::MessageTooLong));
+            }
```

Comment:
I think there is something wrong with this implementation. The test `command_larger_than_max_command_size` fails sporadically on CI (but not on my local machine).

Reply:
In a train right now but I assume that maybe a timeout of a second in not enough. Time is flaky...

Reply:
It's suspicious that it's always this test. I'll dig into it.

Reply:
You are probably right. The test `command_larger_than_max_command_size` loops over different test cases and one of the test cases uses large test data:

```rust
fn command_larger_than_max_command_size() {
    // The server will reject the command because it's larger than the max size
    let max_command_size_tests = [9, 10, 20, 100, 10 * 1024 * 1024];

    for max_command_size in max_command_size_tests {
```

There is another test with large test data, but the other test increases the timeout:

```rust
fn noop_with_large_lines() {
    let mut setup = TestSetup::default();
    // Sending large messages takes some time, especially when running on a slow CI.
    setup.runtime_options.timeout = Some(Duration::from_secs(10));
```

I'll increase the timeout for `command_larger_than_max_command_size`, too.